[JENKINS-56379] Absorb pipeline-model-definition symbol-hetero-list.jelly into hetero-list.jelly

jsoref opened the pull request:
See [JENKINS-56379](https://issues.jenkins-ci.org/browse/JENKINS-56379).

The guts of symbol-hetero-list.jelly becomes:

```
  <f:hetero-list
        name="${name}"
        items="${items}"
        descriptors="${descriptors}"
        addCaption="${addCaption}"
        deleteCaption="${deleteCaption}"
        targetType="${targetType}"
        hasHeader="${hasHeader}"
        oneEach="${oneEach}"
        menuAlign="${menuAlign}"
        honorOrder="${honorOrder}"
        capture="${capture}"
        titleClassMethod="org.jenkinsci.plugins.pipeline.modeldefinition.generator.DirectiveGenerator#getSymbolForDescriptor"
        />
```
### Proposed changelog entries

* Support <f:hetero-list ... titleClassMethod ...> so that symbol-hetero-list.jelly can derive from it

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@abayer

jsoref committed:
[JENKINS-51872] Fix tool dropdown

jsoref committed:
[JENKINS-56379] Absorb pipeline-model-definition symbol-hetero-list.jelly into hetero-list.jelly

Support pipeline-model-definition-plugin: symbol-hetero-list use case

https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/0783180b5d1ed6a398792b37ea46e0aa33ef2b90/pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/symbol-hetero-list.jelly#L126-L129

jsoref commented:
This isn't actually working. I'm confusing my two running instances. But something like this _should_ be possible.

daniel-beck commented on core/src/main/resources/lib/form/hetero-list.jelly:
```diff
@@ -123,10 +127,20 @@ THE SOFTWARE.
+          </j:invokeStatic>
+          <j:set var="title" value="${symbol}: ${descriptor.displayName}"/>
+        </j:if>
+        <j:if test="${titleClassMethod==null}">
```
better done as `when` -- more cumbersome to write (IMO), but less error prone.

oleg-nenashev reviewed (approved):
Formally this is correct.

It might be a slight overengineering to move it to the core, especially since it will take a while to pick up this change in Pipeline Model Definition. But I agree there might be other use-cases which can find the new option useful

vilacides commented:
@abayer @kshultzCB could you please have a look at this PR as pipeline model definition experts? Thanks!

abayer commented:
Seems fine to me - like Oleg said, it'll be a bit before we'd pick this up in Declarative, but this would be worth having.

varyvol commented on core/src/main/resources/lib/form/hetero-list.jelly:
```diff
@@ -123,10 +127,22 @@ THE SOFTWARE.
           <j:set var="capture" value="${attrs.capture?:''}" />
           <local:body deleteCaption="${attrs.deleteCaption}">
-            <l:renderOnDemand tag="tr" clazz="config-page" capture="descriptor,it,instance,${capture}">
+            <l:renderOnDemand tag="tr" clazz="config-page" capture="descriptor,it,instance,h,${capture}">
```
@jsoref what is h?

jvz replied:
`h` is implicitly defined by `initPageVariables()`

jvz replied:
Oh wait, not sure what's going on there. `h` is just a reference to `Functions`.

jsoref replied:
https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/master/pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/symbol-hetero-list.jelly#L132

jsoref commented on core/src/main/resources/lib/form/hetero-list.jelly:
```diff
@@ -123,10 +127,22 @@ THE SOFTWARE.
       <j:set var="instance" value="${null}" />
       <j:set var="descriptors" value="${h.filterDescriptors(it,attrs.descriptors)}" />
```
@varyvol: `h` is used here…

varyvol replied:
Yes, I know it's used here, but why does it need to be added to the other line?

Wadeck commented:
Due to the discussion that started here around the `h`, I investigated a bit more in `pipeline-model-definition` the reason of that "weird" binding. I was especially concerned because the SEC-624 I wrote introduced the regression in that plugin.

<details>

<summary>Technical details</summary>

Quick context, the `h` is a binding for the `Functions` class that provide lots of static helper for Jelly views, this variable is declared during the layout initialization, mainly in `l:layout` and `l:ajax`.

After investigation, the problem was that the [ToolsDirective/config.jelly](https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/281/files#diff-baea812127ddf99065d1d614368368b9R76) was not using the expected `l:ajax` inside the `l:renderOnDemand` and thus, the dynamically loaded part did not have a proper layout (and thus, missing the `h`). Adding the `h` to the binding could be a good solution to this local issue but it's not sustainable in the future because if the core adds something in the layout, that local "hack" will not work any longer (and we will discover the issue only years after, like in this case).
In addition, in this same PR, for consistency, the author also "corrected" another occurences of this `l:renderOnDemand`, and that's concerning us here, because it was in the [symbol-hetero-list.jelly](https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/281/files#diff-055c4d70dcafa085e445ae352b73150dR132). This fix was not required because that one uses the `l:ajax` inside.

In order to avoid future misunderstanding of this complicated framework (Jelly), I proposed [this PR](https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/323) in pipeline-model-definition to simplify the correction and avoid binding the `h`.

</details>

Long story short, it's not wrong to have that `h`, but not useful either and induce other people in error due to that. If you do not mind I will simplify this PR by removing the `h` binding.

Wadeck committed:
Simplify the changes by removing the `h` binding

This variable is provided by the l:ajax tag inside the dynamically loaded fragment

jsoref commented:
@daniel-beck The main reason for this change is to make it possible to switch pipeline from tables to divs at the same time as everyone else, otherwise it will break horribly. And given how painful it is to get plugins released, let alone get things out in any synchronized way, I don't want to imagine having to do it. 

Note: The fact that @oleg-nenashev and @abayer have indicated they'd take a while to adopt this is an indication of the lag I'm worried about. If it would take them a while to pull in this. Imagine how intractable it would be for me to switch tables to divs while trying to dance w/ them?

My wild guess is that w/ this change, they might pull in the change in about 3 months which means I could probably change tables to divs in late August.

w/o this change, I don't think there's actually a way to make it work, one of the key requirements given to me was to not break major plugins. Pipeline is a major plugin.

daniel-beck commented:
I see. That makes sense, and something we need to figure out. Perhaps I'm a bit pessimistic wrt. the dates, but I think we generally don't want to limit Pipeline features to the bleeding edge -- so late 2019 perhaps? Would love to hear from @abayer though, as his plugin would need to actually make use of this component. Otherwise there's no real reason to migrate this control to core.

I wonder whether there's a simple way for plugins to be "forward-compatible", meaning they would query Jenkins for whether it's an updated form overhaul, and if so, switch to a new UI implementation.

jsoref commented:
So, a quick look through my local repos shows this one file as the most interesting ... thing:
https://github.com/jenkinsci/scm-api-plugin/blob/35006747f1293ad2ae5bdb8102693310c956d674/src/main/resources/jenkins/scm/api/form/traits.jelly#L28

oleg-nenashev commented:
Ping @abayer

abayer commented:
This still seems fine by me - we should be able to move to a new enough core after it’s in an LTS

oleg-nenashev commented:
Taking the comment from @abayer and another use-case referenced by @jsoref , I suggest we go forward with the merge
